File: MATLAB/Method/Statistical_methods/Time_domain/LMT/LMT.m
function [Pulse_shape_discrimination_factor] = LMT(Pulse_signal)

%--------------------------------------------------------------------------
% Log Mean Time (LMT) method for pulse shape discrimination.
%
% Discriminates particle types by analyzing the temporal characteristics
% of pulse signals. Computes the natural logarithm of the amplitude-weighted
% mean time to capture differences in pulse decay patterns.
%
% Input:
%   Pulse_signal: Matrix of pulse signals to analyze [n_signals × signal_length]
%
% Output:
%   Pulse_shape_discrimination_factor: Log of mean arrival time
%                                    [n_signals × 1]
%
% Algorithm:
%   1. Create time series t = [1, 2, ..., signal_length]
%   2. For each signal S(t):
%      - Calculate weighted mean time: μt = Σ(t·S(t)) / Σ(S(t))
%      - Compute discrimination factor: ln(μt)
%
% Note:
%   - Returns NaN for signals with zero total amplitude
%   - Time series assumes uniform sampling (1-indexed)
%
% References:
% [1] Dutta, S., et al. "Pulse shape simulation and discrimination using
%     machine learning techniques." Journal of Instrumentation 18.03 (2023):
%     P03038.
% [2] Lee, H. S., et al. "Neutron calibration facility with an Am-Be source
%     for pulse shape discrimination measurement of CsI (Tl) crystals."
%     Journal of Instrumentation 9.11 (2014): P11015.
%--------------------------------------------------------------------------

% Generate time series (1-indexed sample points)
Time_series = 1:size(Pulse_signal, 2);

% Initialize output array
Num_signals = size(Pulse_signal, 1);
Pulse_shape_discrimination_factor = NaN(Num_signals, 1);

% Process each pulse signal
for i = 1:Num_signals
    Signal = Pulse_signal(i, :);
    
    % Calculate amplitude-weighted temporal moments
    Weighted_sum = sum(Signal .* Time_series);  % First moment (Σ t·S(t))
    Total_sum = sum(Signal);                    % Zero moment (Σ S(t))
    
    % Compute log mean time if signal is non-zero
    if Total_sum ~= 0
        Mean_time = Weighted_sum / Total_sum;   % μt = Σ(t·S(t)) / Σ(S(t))
        Pulse_shape_discrimination_factor(i) = log(Mean_time);
    end
    % NaN is maintained for zero-amplitude signals
end

end

File: Method/Statistical_methods/Time_domain/lmt.py
"""
Log Mean Time (LMT) method for pulse shape discrimination.

The method computes discrimination factors by taking the natural logarithm 
of the amplitude-weighted mean time of each pulse signal.

References:
- Lee, H. S., H. Bhang, J. H. Choi, S. Choi, I. S. Hahn, E. J. Jeon, H. W. Joo et al.
  "Neutron calibration facility with an Am-Be source for pulse shape discrimination measurement of CsI (Tl) crystals."
  Journal of Instrumentation 9, no. 11 (2014): P11015.
"""

import numpy as np


def get_psd_factor(
        pulse_signal: np.ndarray
) -> np.ndarray:
    """
    Calculate PSD factors using the Log Mean Time method.
    
    Args:
        pulse_signal: Input signals array of shape (N, L) where N is the number
                     of pulses and L is the length of each pulse
    
    Returns:
        numpy.ndarray: Array of PSD factors for each input pulse, computed as
                      ln(mean_time) where mean_time is the amplitude-weighted
                      average time. NaN values indicate invalid calculations
                      (e.g., zero total amplitude)
    
    Note:
        The time series is 1-indexed, meaning the first sample is at t=1.
    """
    # Create a time series corresponding to each time point in the pulse signals (1-indexed)
    time_series = np.arange(1, pulse_signal.shape[1] + 1)

    num_signals = pulse_signal.shape[0]
    discrimination_factors = np.full(num_signals, np.nan)

    for i, signal in enumerate(pulse_signal):
        # Compute the weighted sum of the pulse amplitudes using the time series
        weighted_sum = np.sum(signal * time_series)
        total_sum = np.sum(signal)

        # Calculate mean time and compute its natural logarithm as the discrimination factor
        if total_sum != 0:
            mean_time = weighted_sum / total_sum
            discrimination_factors[i] = np.log(mean_time)
        else:
            discrimination_factors[i] = np.nan

    return discrimination_factors

